### PR TITLE
Step 1/4 [ObjectDetection]: allow PyCOCOCallback to take apply_nms

### PR DESCRIPTION
**Pull request by tanzhenyu**
The trained result for FasterRCNN got ~40.4 mAP (0.5:0.95), and ~66 mAP (0.5).

inheriting OD Base Model and overriding `predict` only gets to ~28 mAP, which is significantly lower than this PR

Step 2/4: update RetinaNet to disable nms during train and eval, and remove OD Base Model.
Step 3/4: add Batch MultiClass NMS layer that works for TPU
Step 4/4: update FasterRCNN and RetinaNet to take new NMS layer, and set apply_nms to True.

### CODE REVIEW

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 1]
```diff
@@ -225,6 +225,51 @@ def get_config(self):
         return dict(list(base_config.items()) + list(config.items()))
 
 
+class NMSDecoder(tf.keras.layers.Layer):
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/faster_rcnn.py:228]: Should this be a public API?

Would it be possible to use the `non_max_supression` layer for this?

File: keras_cv/callbacks/pycoco_callback.py [hunk 2]
```diff
@@ -19,7 +19,9 @@
 
 
 class PyCOCOCallback(Callback):
-    def __init__(self, validation_data, bounding_box_format, cache=True, **kwargs):
+    def __init__(
+        self, validation_data, bounding_box_format, apply_nms=True, cache=True, **kwargs
```
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/callbacks/pycoco_callback.py:23]: Done

File: keras_cv/callbacks/pycoco_callback.py [hunk 3]
```diff
@@ -19,7 +19,9 @@
 
 
 class PyCOCOCallback(Callback):
```
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/callbacks/pycoco_callback.py:21]: Good point. Done

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 1 above)
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:229]: Added a TODO. For now we don't have a TPU compatible layer, and the one we have is probably buggy given I cannot reproduce a good result

File: examples/models/object_detection/faster_rcnn/train.py [hunk 4]
```diff
@@ -275,6 +282,7 @@ def pad_fn(examples):
     tf.keras.callbacks.TensorBoard(
         log_dir=FLAGS.tensorboard_path, write_steps_per_second=True
     ),
+    PyCOCOCallback(eval_ds, bounding_box_format="yxyx", apply_nms=False),
```
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply examples/models/object_detection/faster_rcnn/train.py:285]: yes, here's the log:
Epoch 18/18
517/517 [==============================] - ETA: 0s - loss: 0.0960 - box_loss: 0.0356 - cls_loss: 0.0403 - rpn_box_loss: 0.0148 - rpn_cls_loss: 0.0053 - mean: 0.0000e+002022-11-29 06:21:48.247378: W tensorflow/core/grappler/optimizers/data/auto_shard.cc:547] The `assert_cardinality` transformation is currently not handled by the auto-shard rewrite and will be removed.
154/154 [==============================] - 16s 100ms/step
creating index...
index created!
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=5.56s).
Accumulating evaluation results...
DONE (t=0.86s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.406
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.666
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.446
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.037
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.213
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.487
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.358
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.469
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.469
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.047
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.263
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.558

File: keras_cv/callbacks/pycoco_callback.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/callbacks/pycoco_callback.py:23]: I think it's worth considering inverting the meaning of this parameter (e.g. `apply_nms=True` means that the callback will apply NMS) or renaming it to something like `model_performs_nms`

**Review by ianstenbit — APPROVED** [review]
Thanks!
